core/src/main/java/org/infinispan/commands/statetransfer/AbstractStateTransferCommand.java

```diff
@@ -0,0 +1,78 @@
+package org.infinispan.commands.statetransfer;
+
+import java.io.IOException;
+import java.io.ObjectInput;
+import java.io.ObjectOutput;
+
+import org.infinispan.commands.TopologyAffectedCommand;
+import org.infinispan.commands.remote.BaseRpcCommand;
+import org.infinispan.commons.util.IntSet;
+import org.infinispan.remoting.transport.Address;
+import org.infinispan.util.ByteString;
+
+/**
+ * Base class for commands related to state transfer.
+ *
+ * @author Ryan Emerson
+ * @since 11.0
+ */
+abstract class AbstractStateTransferCommand extends BaseRpcCommand implements TopologyAffectedCommand {
+
+   private byte commandId;
+   protected int topologyId;
+   protected IntSet segments;
+
+   AbstractStateTransferCommand(byte commandId, ByteString cacheName) {
+      super(cacheName);
+      this.commandId = commandId;
+   }
+
+   AbstractStateTransferCommand(byte commandId, ByteString cacheName, Address origin, int topologyId, IntSet segments) {
+      this(commandId, cacheName);
+      this.origin = origin;
+      this.topologyId = topologyId;
+      this.segments = segments;
+   }
+
+   @Override
+   public boolean isReturnValueExpected() {
+      return true;
+   }
+
+   @Override
+   public boolean canBlock() {
+      // All state request commands need to wait for the proper topology
+      return true;
+   }
+
+   @Override
+   public int getTopologyId() {
+      return topologyId;
+   }
+
+   @Override
+   public void setTopologyId(int topologyId) {
+      this.topologyId = topologyId;
+   }
+
+   public IntSet getSegments() {
+      return segments;
+   }
+
+   @Override
+   public byte getCommandId() {
+      return commandId;
+   }
+
+   @Override
+   public void writeTo(ObjectOutput output) throws IOException {
+      output.writeObject(origin);
+      output.writeObject(segments);
```

Comment:
what about using `IntSetsExternalization.writeTo()` and `IntSetsExternalization.readFrom()`?